### PR TITLE
feat: Allow blocking until primary rate limit is reset

### PR DESCRIPTION
**Pull request by erezrokah**
Fixes https://github.com/google/go-github/issues/3114

Feedback very much welcomed, some notes:
1. I used the same approach as `bypassRateLimitCheck` with the exception that `bypassRateLimitCheck` is private, not sure  if to use another dedicated exported `requestContext` type or also make `bypassRateLimitCheck` public for consistency
2. Checking `req.Context().Value(SleepUntilPrimaryRateLimitResetWhenRateLimited) == true` is also an option (instead of `!=nil`) and might be more readable, but that is not consistent with the `bypassRateLimitCheck` check
3. A possible improvement can be to add a log message/invoke a callback to consumers so they can show an indication the sleep is happening to avoid looking like something is stuck. If that makes sense it should be in a separate PR I think as that's would be a new API/interface for this module
4. If this approach makes send I will add tests for it

### CODE REVIEW

File: github/github.go
Comment (gmlewis) [inline github/github.go:897]: I think this is probably OK, but am wondering about two things:
* this recursively calls c.BareDo - at what point should it truly fail?
* The defered body close on line 876 doesn't actually happen until the recursion fully unwraps

Again, I think this is probably all fine.
Did you try testing this out, perchance?
I know the whole point is to prevent rate limit violations - but it might be nice to verify that this does what we want.

**Comment by gmlewis** [comment]
It would also be nice to add one or more unit tests that would increase the code coverage for the new code.

File: github/github.go
Comment (erezrokah) [reply github/github.go:897]: I did test it and the retry works. Good point on the defer let me see if I can find a good solution for it. Another thing I'm considering is to only retry once

**Comment by erezrokah** [comment]
> Thank you, @erezrokah ! LGTM.
> 
> Awaiting second LGTM+Approval from any other contributor to this repo before merging.

Thanks for the quick review and feedback, I also added a 1 second buffer to the sleep time to avoid any edge cases

**Review by gmlewis — APPROVED** [review]
Thanks, @erezrokah .

Still awaiting second LGTM+Approval from any other contributor to this repo before merging.

All contributions, including code reviews, by other volunteers is greatly appreciated, as we have several outstanding PRs that need reviews.  Thanks.

**Comment by erezrokah** [comment]
Hi @gmlewis I'm wondering if I should update the PR to take into account context cancellation, by changing `sleepUntilResetWithBuffer` to

```go
timer := time.NewTimer(time.Until(reset) + buffer)
select {
case <-ctx.Done():
	if !timer.Stop() {
	<-timer.C
	}
	return nil, ctx.Err()
case <-timer.C:
}
```

**Comment by gmlewis** [comment]
Ah, that's an excellent idea, @erezrokah - I think you are right that this should be added.

**Comment by erezrokah** [comment]
> Ah, that's an excellent idea, @erezrokah - I think you are right that this should be added.

Added https://github.com/google/go-github/pull/3117/commits/22eba4c83515649fe06e82b4b7dad938a06b5044. In one case I had to wrap the context error in a rate limit error. I can also refactor the code a bit to have `checkRateLimitBeforeDo` return an `error` instead of `*RateLimitError`, but thought it would be better to get some feedback first

File: github/github.go
Comment (gmlewis) [inline github/github.go:967]:
```suggestion
			if err := sleepUntilResetWithBuffer(req.Context(), rate.Reset.Time); err == nil {
```

File: github/github.go
Comment (gmlewis) [inline github/github.go:896]:
```suggestion
			if err := sleepUntilResetWithBuffer(req.Context(), rateLimitError.Rate.Reset.Time); err != nil {
```

**Review by gmlewis — COMMENTED** [review]
Just a couple nits, otherwise LGTM.

**Review by gmlewis — APPROVED** [review]
Thank you, @erezrokah !
LGTM.

**Review by be0x74a — APPROVED** [review]
LGTM!

**Comment by gmlewis** [comment]
Thank you, @be0x74a !
Merging.

**Comment by erezrokah** [comment]
Thanks for the review(s) @be0x74a and @gmlewis, when can we expect a new release to be tagged with this commit?

**Comment by gmlewis** [comment]
> Thanks for the review(s) @be0x74a and @gmlewis, when can we expect a new release to be tagged with this commit?

I'm away from Internet this week so the earliest will be this Friday.

**Comment by gmlewis** [comment]
This is now available here: https://github.com/google/go-github/releases/tag/v62.0.0

**Pull request merged**
